File: 0316-remove-duplicate-letters/0316-remove-duplicate-letters.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
    // output must be a subsequence
        int len = s.size();
        string res = "";
        // lastIndex for storing the lastIndex of the char
        vector<int> lastIndex(26, 0);
        vector<bool> taken(26, false);
        stack<char> st;

        for(int i=0; i<len; i++){
            int value = s[i] - 'a';
            lastIndex[value] = i;
        }

        for(int i=0; i<len; i++){
            int curr = s[i] - 'a'; // stores the integer value of char

            if(taken[curr]) // already taken
                continue;

            // check last index of the top value of stack
            while(!st.empty() && st.top() > s[i] && lastIndex[st.top() - 'a'] > i){
                // if(st.top() > s[i] && lastIndex[st.top() - 'a'] > i){ // c > b
                    taken[st.top() - 'a'] = false;
                    st.pop();
                // }
            }

            st.push(s[i]);
            taken[curr] = true;
        }

        while(!st.empty()){
            res += st.top();
            st.pop();
        }

        cout << res;
        reverse(res.begin(), res.end());
        return res;
    }
};

File: 0316-remove-duplicate-letters/0316-remove-duplicate-letters.java
class Solution {
    public String removeDuplicateLetters(String s) {
        int len = s.length();
        String res = "";
        int[] lastIndex = new int[26]; // can also declare int[] arr;
        Arrays.fill(lastIndex, 0);

        boolean[] taken = new boolean[26];
        Arrays.fill(taken, false);

        Stack<Character> st = new Stack<Character>();

        for(int i=0; i<len; i++){
            int value = s.charAt(i) - 'a';
            lastIndex[value] = i;
        }

        for(int i=0; i<len; i++){
            int curr = s.charAt(i) - 'a'; // stores the integer value of char

            if(taken[curr]) // already taken
                continue;

            // arrange stack like maxheap
            // check last index of the top value of stack
            while(!st.isEmpty() && st.peek() > s.charAt(i) && lastIndex[st.peek() - 'a'] > i){
                    taken[st.peek() - 'a'] = false;
                    st.pop();
            }

            st.push(s.charAt(i));
            taken[curr] = true;
        }

        while(!st.empty()){
            res += st.peek();
            st.pop();
        }

        System.out.println(res);
        StringBuilder sb = new StringBuilder();
        sb.append(res);
        sb.reverse();

        return sb.toString();
    }
}